walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,156 @@
+      if (hasOptionalAttendees && !hasAllAttendees(event, request)) {
+        continue;
+      }
+      if (!hasOptionalAttendees && !hasRequiredAttendees(event, request)) {
+        continue;
+      }
+
+      for (TimeRange slot : slots) {
+        TimeRange before = TimeRange.fromStartEnd(slot.start(), eventTime.start(), false);
+        TimeRange after = TimeRange.fromStartEnd(eventTime.end(), slot.end() - 1, true);
+
+        if (slot.contains(eventTime)) {
+          removeFromMiddle(curSlots, request.getDuration(), before, after, slot);
+          break;
+        } else if (slot.contains(eventTime.end())) {
+          removeFromEnd(curSlots, request.getDuration(), after, slot);
+          break;
+        } else if (slot.contains(eventTime.start())) {
+          removeFromStart(curSlots, request.getDuration(), before, slot);
+          break;
+        }
+      }
+    }
+    return curSlots;
+  }
+
+  /*
+   * Replaces slot with two new slots, with a gap in the middle corresponding
+   * to the space that the event takes (and thus the meeting cannot take place in).
+   */
+  private void removeFromMiddle(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange before, TimeRange after, TimeRange slot) {
+    if (before.duration() >= reqDuration) {
+      curSlots.add(before);
+    }
+    if (after.duration() >= reqDuration) {
+      curSlots.add(after);
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * Removes everything in the slot before TimeRange after, corresponding to the space
+   * that the event takes (and thus the meeting cnanot take place in).
+   */
+  private void removeFromEnd(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange after, TimeRange slot) {
+    if (after.duration() >= reqDuration) {
+      curSlots.add(after);
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * Removes everything in the slot after TimeRange before, corresponding to the space
+   * that the event takes (and thus the meeting cnanot take place in).
+   */
+  private void removeFromStart(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange before, TimeRange slot) {
+    if (before.duration() >= reqDuration) {
+      curSlots.add(before);
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * Return whether attendees for event overlap with required attendees for request
+   * (i.e. at least one attendee for request is an attendee for event).
+   */
+  private boolean hasRequiredAttendees(Event event, MeetingRequest request) {
+    for (String reqAttendee : request.getAttendees()) {
+      if (event.getAttendees().contains(reqAttendee)) {
+          return true;
+      }
+    }
+    return false;
+  }
+
+  /*
+   * Return whether attendees for event overlap with all (required 
+   * and optional) attendees for request (i.e. at least one attendee 
+   * for request is an attendee for event).
+   */
+  private boolean hasAllAttendees(Event event, MeetingRequest request) {
```

Comment:
similar comment as above - perhaps you can just check the opposite and/or update the naming to better reflect this logic.

Reply:
Also, optionally, consider replacing these with a method that checks whether there's any overlap in two groups of attendees (represented by Collections of Strings) and pass different groups of attendees to it for the different checks. e.g.
```
private static boolean hasAnyInCommon(Collection<String> groupA, Collection<String> groupB) {
  // TODO
}

// inside findTimeslots()...
if (hasOptionalAttendees && !hasAnyInCommon(event.getAttendees(), request.getAttendees()) && !hasAnyInCommon(event.getAttendees(), request.getOptionalAttendees()) {
  continue;
}
if (!hasOptionalAttendees && !hasAnyInCommon(event.getAttendees(), request.getAttendees())) {
  continue;
}
// ...end excerpt from findTimeslots()

// or, another way to do the logic in findTimeslots()...
if (!hasAnyInCommon(event.getAttendees(), request.getAttendees()) {
  if (!hasOptionalAttendees || !hasAnyInCommon(event.getAttendees(), request.getOptionalAttendees()) {
    continue;
  }
}
// ...end excerpt from findTimeslots()

Reply:
thanks so much -- I think this would be a lot cleaner/more intuitive to read.